File: c/perfect-numbers/perfect_numbers.c
#include "perfect_numbers.h"
#include <stddef.h>

kind classify_number(int number) {
  if (number < 1) {
    return ERROR;
  }

  if (number == 1) {
    return DEFICIENT_NUMBER;
  }

  int aliquot_sum = number == 1 ? 0 : 1;

  size_t divisor = 2;
  size_t quotient = number / divisor;
  size_t remainder = number % divisor;

  while (quotient >= divisor) {
    if (remainder == 0) {
      aliquot_sum += divisor;

      if (quotient != divisor) {
        aliquot_sum += quotient;
      }

      if (aliquot_sum > number) {
        return ABUNDANT_NUMBER;
      }
    }

    divisor++;
    quotient = number / divisor;
    remainder = number % divisor;
  }

  if (aliquot_sum < number) {
    return DEFICIENT_NUMBER;
  }

  return PERFECT_NUMBER;
}

File: python/perfect-numbers/perfect_numbers.py
def calculate_factors(dividend: int):
    if dividend != 1:
        yield 1

    divisor = 2
    quotient, remainder = divmod(dividend, divisor)

    while quotient >= divisor:
        if remainder == 0:
            yield divisor

            if quotient != divisor:
                yield quotient

        divisor += 1
        quotient, remainder = divmod(dividend, divisor)


def classify(number: int):
    """A perfect number equals the sum of its positive divisors.

    :param number: int a positive integer
    :return: str the classification of the input integer
    """

    if number < 1:
        raise ValueError("Classification is only possible for positive integers.")

    aliquot_sum = sum(calculate_factors(number))

    if number < aliquot_sum:
        return "abundant"

    if number > aliquot_sum:
        return "deficient"

    return "perfect"
